File: arrays/maximizeDistanceToClosestPerson.java
package arrays;
// https://leetcode.com/problems/maximize-distance-to-closest-person/
// Solution : https://www.youtube.com/watch?v=Ngf1XUDfK7M4
public class maximizeDistanceToClosestPerson {
    public static void main(String[] args) {

        SolutionMDTCP solution = new SolutionMDTCP();
        System.out.println(solution.maxDistToClosest(new int[]{1,0,0,0,1,0,1}));
        System.out.println(solution.maxDistToClosest(new int[]{0,0,0,0,1,0,1}));
        System.out.println(solution.maxDistToClosest(new int[]{1,0,0,0,0,0}));

        System.out.println(solution.maxDistToClosestApproach2(new int[]{1,0,0,0,1,0,1}));
        System.out.println(solution.maxDistToClosestApproach2(new int[]{0,0,0,0,1,0,1}));
        System.out.println(solution.maxDistToClosestApproach2(new int[]{1,0,0,0,0,0}));   
    }
    
}

class SolutionMDTCP{
    public int maxDistToClosest(int[] seats) {

        int len = seats.length;
        int gap = Integer.MIN_VALUE;
        int lst = 0;
        int fst = 0;
        int ind = 0;
        int steps = 0;

        if (seats[0] == 0) {

            for (int i = 0; i < len; i++) {
                if (seats[i] == 1) {
                    ind = i;
                    fst = steps;
                    break;
                } else
                    steps++;
            }
        }

        steps = 0;
        for (int i = ind; i < len; i++) {

            if (i == len - 1 && seats[i] == 0) {
                lst = steps + 1;
                break;
            }

            if (seats[i] == 0) {
                steps++;
            }
            if (seats[i] == 1) {
                if (steps > gap)
                    gap = steps;
                steps = 0;
            }
        }

        int mid_gap = (int) Math.floor((gap + 1) / (double) 2);

        if (fst >= lst && fst >= mid_gap)
            return fst;
        else if (mid_gap >= fst && mid_gap >= lst)
            return mid_gap;
        else
            return lst;

    }

    // TC: O(N)
    // SC: O(1)

    // Easy and Efficient Solution

    // Runtime: 2 ms, faster than 77.87% of Java online submissions for Maximize Distance to Closest Person.
    // Memory Usage: 46.6 MB, less than 15.31% of Java online submissions for Maximize Distance to Closest Person.
    
    public int maxDistToClosestApproach2(int[] seats) {
        
        int totalSeats = seats.length;
        int firstOccurenceOfOne = -1;
        int countOfZeroesBetweenTwoSeats=0;
        int maximumDistance = 0;
        
        for (int i = 0; i<totalSeats; i++){
            if (seats[i]==0) countOfZeroesBetweenTwoSeats++;
            else if (seats[i]==1){
                if(firstOccurenceOfOne ==-1){
                    maximumDistance = countOfZeroesBetweenTwoSeats;
                    firstOccurenceOfOne =i;
                }else{
                    maximumDistance= Math.max(maximumDistance,(countOfZeroesBetweenTwoSeats+1)/2);
                }
                countOfZeroesBetweenTwoSeats=0;
            }
            
        }
        return Math.max(maximumDistance,countOfZeroesBetweenTwoSeats);
    }
}

File: arrays/maximizeDistanceToClosestPerson.py
# https://leetcode.com/problems/maximize-distance-to-closest-person/
# Solution : https://www.youtube.com/watch?v=Ngf1XUDfK7M

class Solution:
    
    # TC : O(N)
    # SC : O(N)
    def maxDistToClosest(self, seats):
        """
        :type seats: List[int]
        :rtype: int
        """
        
        n = len(seats)
        left = [n]*n
        right = [n]*n
        
        
        for i in range(n):
            if seats[i]==1:
                left[i]=0
            elif i>0 :
                left[i]=left[i-1]+1
              
        
        for i in range(n-1, -1,-1):
            if seats[i]==1:
                right[i]=0
            elif i<n-1:
                right[i]=right[i+1]+1
    
        
        maximumDistance =float('-inf')
        for i in range(n):
            minimumDistanceBetweenRightAndLeftOne = min(left[i],right[i])
            maximumDistance = max(minimumDistanceBetweenRightAndLeftOne, maximumDistance)
            
        return maximumDistance
    
    
    # TC : O(N)
    # SC : O(1)
    
    # Runtime: 108 ms, faster than 84.75% of Python online submissions for Maximize Distance to Closest Person.
    # Memory Usage: 14.6 MB, less than 33.90% of Python online submissions for Maximize Distance to Closest Person.
    
    def maxDistToClosestApproach2(self, seats):
        """
        :type seats: List[int]
        :rtype: int
        """
        seatsAvailable = len(seats)
        firstOccupiedSeat = -1
        countOfZeroBetweenTwoSeats = 0
        maximumDistance = float('-inf')
        
        for i in range(seatsAvailable):
            
            if seats[i]==0:
                countOfZeroBetweenTwoSeats+=1
            else:
                if firstOccupiedSeat == -1 :
                    maximumDistance = countOfZeroBetweenTwoSeats
                    firstOccupiedSeat =i
                else :
                    maximumDistance = max(maximumDistance,(countOfZeroBetweenTwoSeats+1)//2)
                countOfZeroBetweenTwoSeats=0
                    
        return max(maximumDistance, countOfZeroBetweenTwoSeats)
    
        '''
        Example : 
        
        1 0 0 0 0 1 0
        
        We record the first occurence of one and count the zeros till next occurence 
        We find the middle position in between next occurence and note maximum in it
        
        edge cases : 
        
        0,0,0,0,1
        Till we find our first occurence of one we substitue the count in maximum distance
        Makes sense right
        
        1,0,0,0,0,0
        At last we check the maximum value so that we can get rightmost seat
        
        '''